google/cloud/bigquery/table.py

```diff
@@ -452,17 +453,20 @@ def schema(self):
                 instance or a compatible mapping representation of the field.
         """
         prop = self._properties.get(self._PROPERTY_TO_API_FIELD["schema"])
-        if not prop:
+        if not prop:  # if empty Schema, empty list, None
             return []
```

Comment:
Could we return an "empty" Schema here for consistency?

```suggestion
            return Schema()
```

Reply:
Along the lines of least surprise... if the user passes in an empty `list` or `None`, they will get back an empty `list`. If they pass in an empty `Schema()` they will get back an empty `Schema()`

Reply:
@tswast 

Does the following influence your thoughts on this comment?

Historically, we have returned an empty `list`.
An existing user would still expect a `list`.
A new user should probably not see an empty `Schema` object, since they are only needed if the user is gonna try to access an external data source (where they have to identify the `foreign_type_info`).

My general principal, focused on backwards compatibility is:
* avoid as many changes to what the code did before, unless necessary
* as much as possible, force the choice to move to Schema to be an explicit decision by any user (especially since external data sources are likely to be the exception, not the rule).

Reply:
I could go either way on this one. One one hand, using `Schema` everywhere would make the new Schema class more discoverable and fewer cases to consider when type checking. On the other hand, providing a real list in all cases where `Schema` is not needed would provide the most backwards compatible option.

On further consideration, I agree with your assessment that avoiding breaking existing users is preferable. +1 on empty list in this case.

Reply:
@tswast 

this is a related concept.
Right now, if someone gives me a Schema object, I have been storing that in _properties and I see your point about avoiding that.

That complicates getting a schema object back out, if/when you need it.

What are your thoughts on this (I seem to recall you mention keeping objects in a separate place and ensuring that _properties is a view into the object)

```

    @schema.setter
    def schema(self, value):
        api_field = self._PROPERTY_TO_API_FIELD["schema"]

        if value is None:
            self._properties[api_field] = None
        elif isinstance(value, Schema):
            # store original as selt.<original_schema_object> # TODO find a better name
            self.<original_schema_object> = value
            self.properties[api_field] = self.<original_schema_object>.to_api_repr()
        else:
            value = _to_schema_fields(value)
            self._properties[api_field] = {"fields": _build_schema_resource(value)}


    @property
    def schema(self):
        # if user stored a Schema object, short circuit and return early.
        <original_schema_object> = self.get(<original_schema_object>, None)
        if isinstance(original_schema_object, Schema):
            return self.<original_schema_object>

        # if user did not store a Schema object, behave as normal
        prop = self._properties.get(self._PROPERTY_TO_API_FIELD["schema"])
        if not prop: # either None OR empty list  
            return []
        
        return _parse_schema_resource(prop)
```